### PR TITLE
Resume instance by sending messages from self

### LINKED ISSUE #679: Handle resumption/rebroadcast via WAL
Now that there is WAL, ideally a participant should be able to resume progress within a GPBFT instance if restarted. To do so the participant only requires to rebroadcast messages that it would have done prior to restart (i.e. `broadcastState`). If all participants do this then it should be enough to resume participation without any further state-keeping or bespoke steps after restart.

### PR DESCRIPTION
**Pull request by masih**
WAL keeps track of all the messages and we cache messages from the latest instance for rebroadcast purposes.

Before starting a new instance, if there are any messages from self for that instance feed them to the participant. That should speed up catch up at that round and effectively aid resume progress for the participant.

Fixes #679

### CODE REVIEW

File: host.go [hunk 1]
```diff
@@ -289,6 +289,28 @@ func (h *gpbftRunner) receiveCertificate(c *certs.FinalityCertificate) error {
 	log.Infow("skipping forwards based on cert", "from", currentInstance, "to", nextInstance)
 
 	nextInstanceStart := h.computeNextInstanceStart(c)
+
+	// Look for any existing messages in WAL for the next instance, and if there is
```
Comment (Stebalien) [inline host.go:293]: Hm. I guess technically we'll see this on start. Although we'll never see it for instance 0.

Any reason not to replay everything immediately in `gpbftRunner.Start`?

Comment (Stebalien) [reply host.go:293]: Or we can look at the current instance at that point.

Comment (masih) [reply host.go:296]: That would take more gymnastics to get the current instance 🤷 

What I have done in the latest changes is to refactor calls to participant's `StartInstanceAt`, and replay message for the instance being started every time. 

Please let know what you think.

Comment (Stebalien) [reply host.go:296]: Oh, that's much nicer.

**Review by Stebalien — APPROVED** [review]

File: host.go [hunk 2]
```diff
@@ -289,7 +289,32 @@ func (h *gpbftRunner) receiveCertificate(c *certs.FinalityCertificate) error {
...
+		for _, message := range messages {
+			replay = append(replay, message...)
+		}
+	}
+	h.msgsMutex.Unlock()
+	for _, message := range replay {
+		if validated, err := h.participant.ValidateMessage(message); err != nil {
+			log.Warnw("invalid self message", "message", message, "err", err)
+		} else if err := h.participant.ReceiveMessage(validated); err != nil {
+			log.Warnw("failed to send resumption message", "message", message, "err", err)
```
Comment (Kubuxu) [inline host.go:314]: The only thing I'm worried is the ordering of messages but it is probably fine.

**Review by Kubuxu — APPROVED** [review]

File: host.go (hunk 2 above)
Comment (masih) [reply host.go:314]: It should just work, but no harm in sorting them for a more optimal resumption 👍

Comment (masih) [reply host.go:340]: OK added

**Review by Kubuxu — APPROVED** [review]

**Pull request merged**